### ISSUE TITLE
Curiosity about grapheme training

### ISSUE THREAD

**Issue by drdhaval2785** [flame-cai/win64-local-ocr-tool#8]
Currently the segmentation seems to extract bounding boxes for all graphemes. Then they are connected (maybe by taking midpoints) by connecting them as graph nodes and then line images are generated for connected characters, by masking away area other than grapheme.

Would it be worth a try to train  model to recognise at grapheme level? We can use the output from grapheme training to generate sentences, as we already have graphs connecting the nodes.

Why I feel that it will be useful linguistically.

1. Currently the results I get are very encouraging. It makes mistakes on graphemes which are very similar like घ/थ. द्व/द्ध etc. 
2. It sometimes adds new graphemes even if there are no graphemes visually e.g. इरति instead of इति.
3. The writing style of manuscripts use the same (or visually) undifferentiable grapheme for प or य as the second letter of a conjoint. E.g. tpu and tyu are the same grapheme. When we train our model with correct text, we are making the machine confused. It then makes random guesses about y and p.

**Comment by kartikchincholikar** [comment]
@drdhaval2785 this is a very interesting idea. 

In the literature review we observed the field to be moving away from character-by-character recognition, to prefer full text line recognition. However perhaps grapheme level image2text recognition might work better than character level one.


Quoting text from literature review:

> Segmenting lines and words into isolated characters and sub-characters before recognition,
> however, is a common source of error when digitising Devanāgarī text because of the complex
> nature of the Devanāgari script, which contains more than 800 possible character shapes, which
> are made up of combinations of a base character set of 127. Each of the consonant-consonant and consonant-vowel combinations takes different forms based on their position in the word (Karayil
> et al., 2015). To tackle this issue, Sankaran et al. (2013) and Karayil et al. (2015) built on the
> work of Graves et al. (2007) and removed the need to segment document images into isolated
> words and characters using BiLSTMs (Bi-Directional Long Short Term Memory Networks) and
> CTC (Connectionist Temporal Classification) loss. In these papers, the neural network takes
> an image of a line of printed text as input and predicts a sequence of Devanāgari Unicode code
> points as its output.

@ShagunDwivedi

**Comment by drdhaval2785** [comment]
One more curiosity would be
Would grapheme level cost function value give some idea about the confidence of the machine in classifying the said grapheme?
If so, we can highlight the places where machine has 'above threshold' cost value. 
Threshold may be identified statistically by seeing the distribution of the cost function over all graphemes identified.

**Comment by ShagunDwivedi** [comment]
The recognition model is a CNN-BiLSTM-CTC trained by EasyOCR on printed Devanagari characters and then finetuned by us on different Sanskrit manuscripts. The CNN performs feature extraction on the image to generate feature maps, and to quote [this paper](https://arxiv.org/abs/1507.05717), 

> Each column in the resulting feature map by a feature extractor has a corresponding distinguishable receptive field along the horizontal line of the input image. These features are used to estimate the character on each receptive field.

What this means is that mostly a single grapheme, or even an English alphabet for that matter, may have multiple columns assigned to it in the feature map. The BiLSTM then maps the visual feature map to a contextual feature map, which can also be called sequence labeling (For इति, this might look like इ इ र ि  ि त, where the model mistook second half of the इ for र). Then, the CTC model then decodes the sequence ([this](https://distill.pub/2017/ctc/) would be a good read to understand what CTC does) to get इरति.

However, I think a different model architecture which can perform grapheme-level segmentation with precision will surely reduce these errors.
